### PR TITLE
Add layer to Simulator interfaces to allow other step by step simulations

### PR DESCRIPTION
**Pull request by dabacon**
This will allow the density matrix and mixture simulators to work with the same interface for the top level interfaces.

Some notes:
* Change in the contract for the simulator iterator and empty circuits.  This eliminates so gnarly handling of this case.
* The type annotations had to be reduce to Iterator from Iterator[StepResult] because Iterator is covariant.  
* This now allows both StepResult and TrialResult for wave function simulations to implement the StateVectorMixin (which has now been made into a proper mixin).
* Mypy doesn't like init methods for mixins.

### CODE REVIEW

File: cirq/__init__.py [hunk 1]
```diff
@@ -197,21 +197,26 @@
...
     sample_state_vector,
+    SimulatesSamples,
+    SimulatesFinalState,
+    SimulatesIntermediateState,
+    SimulatesIntermediateWaveFunction,
+    SimulationTrialResult,
+    Simulator,
+    SparseSimulatorStep,
+    StateVectorMixin,
+    StateVectorMixin,
```
Comment (Strilanc) [inline cirq/__init__.py]: Double import

File: cirq/sim/simulator.py [hunk 2]
```diff
@@ -206,17 +212,15 @@ def simulate(
...
-            initial_state: If an int, the state is set to the computational
-                basis state corresponding to this state. Otherwise  if this
-                is a np.ndarray it is the full initial state. In this case it
-                must be the correct size, be normalized (an L2 norm of 1), and
-                be safely castable to an appropriate dtype for the simulator.
+            qubit_order: Determines the canonical ordering of the qubits. This
+                is often used in specifying the initial state, i.e. the
+                ordering of the computational basis states.
+            initial_state: The initial state for the simulation. The  form of
+                this state depends on the simulation implementation.  See
```
Comment (Strilanc) [inline cirq/sim/simulator.py]: Could this state dependence be packaged into a generic parameter on the class? If it causes an explosion of generics everywhere, probably best not to for now.

File: cirq/sim/simulator_test.py [hunk 3]
```diff
@@ -90,39 +96,21 @@ def steps(*args, **kwargs):
     np.testing.assert_equal(result.measurements['b'], [True, False])
     assert set(result.measurements.keys()) == {'a', 'b'}
     assert result.params == param_resolver
-    np.testing.assert_equal(result.final_state, final_state)
+    print(result)
```
Comment (Strilanc) [inline cirq/sim/simulator_test.py]: print

File: cirq/sim/simulator_test.py [hunk 4]
```diff
@@ -135,170 +123,26 @@ def steps(*args, **kwargs):
                                        qubit_order=qubit_order,
                                        initial_state=2)
     expected_results = [
-        cirq.SimulationTrialResult(
+        cirq.sim.simulator.SimulationTrialResult(
```
Comment (Strilanc) [inline cirq/sim/simulator_test.py]: Should just be `cirq.SimulationTrialResult`.

File: cirq/__init__.py (hunk 1 above)
Comment (dabacon) [reply cirq/__init__.py]: fixed

File: cirq/sim/simulator_test.py (hunk 3 above)
Comment (dabacon) [reply cirq/sim/simulator_test.py]: removed

File: cirq/sim/simulator_test.py (hunk 4 above)
Comment (dabacon) [reply cirq/sim/simulator_test.py]: done and cleaned up some imports

File: cirq/sim/simulator.py (hunk 2 above)
Comment (dabacon) [reply cirq/sim/simulator.py]: this is probably the worst part of this interface right now.  In particular note the qubit order, which is not actually always necessary for some simulators.   I agree we should try to find a better solution, maybe after we get new simulators in this will become more clear.

**Comment by dabacon** [comment]
Had to add annotation for repr tests to avoid running them in python 2 due to unicode translation.
